dokka-runners/dokka-gradle-plugin/src/test/kotlin/utils/ShouldBeADirectoryWithSameContentAsTest.kt

```diff
@@ -0,0 +1,145 @@
+
+        failure.shouldHaveMessage("expectedDir '$expectedDir' is not a directory (exists:false, file:false)")
+    }
+
+    test("when actual directory doesn't exist, expect failure") {
+        val expectedDir = tempDir()
+        val actualDir = tempDir()
+
+        actualDir.deleteRecursively()
+
+        val failure = shouldFail { expectedDir.shouldBeADirectoryWithSameContentAs(actualDir) }
+
+        failure.shouldHaveMessage("actualDir '$actualDir' is not a directory (exists:false, file:false)")
+    }
+
+    test("when directories have different files, expect failure") {
+        val expectedDir = tempDir().apply {
+            resolve("file0.txt").writeText("valid file")
+            resolve("file1.txt").writeText("not in actual")
+            resolve("file2.txt").writeText("not in actual")
+            resolve("file3.txt").writeText("not in actual")
+        }
+
+        val actualDir = tempDir().apply {
+            resolve("file0.txt").writeText("valid file")
+            resolve("file-a.txt").writeText("not in expected")
+            resolve("file-b.txt").writeText("not in expected")
+            resolve("file-c.txt").writeText("not in expected")
+        }
+
+        val failure = shouldFail { expectedDir.shouldBeADirectoryWithSameContentAs(actualDir) }
+
+        failure.shouldHaveMessage(
+            """
+            actualDir is missing 3 files:
+              - file1.txt
+              - file2.txt
+              - file3.txt
+            actualDir has 3 unexpected files:
+              - file-a.txt
+              - file-b.txt
+              - file-c.txt
+            """.trimIndent()
+        )
+    }
+
+    test("when file in directories has different content, expect failure with contents of file") {
+        val expectedDir = tempDir().apply {
+            resolve("file.txt").writeText("content")
+        }
+
+        val actualDir = tempDir().apply {
+            resolve("file.txt").writeText("unexpected content")
+        }
+
+        val failure = shouldFail { expectedDir.shouldBeADirectoryWithSameContentAs(actualDir) }
+
+        failure.shouldHaveMessage(
+            """
+            file.txt has 1 differences in content:
+                --- file.txt
+                +++ file.txt
+                @@ -1,1 +1,1 @@
+                -content
+                +unexpected content
+            """.trimIndent()
+        )
+    }
+
+    test("when binary file in directories has different content, expect failure with contents of file") {
+        val expectedDir = tempDir().apply {
+            resolve("file.bin").writeBytes(PNG_BYTES)
+        }
+
+        val actualDir = tempDir().apply {
+            resolve("file.bin").writeBytes(PNG_BYTES.reversedArray())
+        }
+
+        val failure = shouldFail { expectedDir.shouldBeADirectoryWithSameContentAs(actualDir) }
+
+        failure.shouldHaveMessage(
+            """
+            file.bin has 2 differences in content:
+                --- file.bin
```

Comment:
TBH, it would be better to just skip binary files difference printing if possible and check by hash/size only, it's not really useful in my opinion and diff doesn't really helps here.

E.g if we will change the image in our frontend, not sure that diff will help somehow :)

Reply:
I agree, but as far as I can tell JVM doesn't have a reliable way of determining if a file is a binary file or not. If I think about our priorities, this should be good enough for now.

What do you think?

Reply:
Agree, the only reason I've mentioned it, is because we have custom logic for this, e.g `readByteLines`.

Reply:
I figured out an improvement: just catch IOExceptions when reading the content, and print more details otherwise. So now binary file comparison failures look like this:


```
file.bin has 1 differences in content:
    --- file.bin
    +++ file.bin
    @@ -1,4 +1,1 @@
    -Failed to read file content
    -java.nio.charset.MalformedInputException Input length = 1
    -file size: 100
    -checksum: c+g2IzgALWMi8irrW5xr/gB32XVU8WTtaQ8hkD8qXzE=
    +This file in the actual dir has valid UTF-8 content
```

wdyt?

Reply:
yeah, looks better!
minor: probably it would be better to leave binary contents for both expect and actual values as it was before, as this makes more sense I think

Reply:
Cool, thanks.

The example used in the test is a bit atypical. In real life both files are likely to be binaries, so the logged message will be different.